core/unix/memquery_test.h

```diff
@@ -0,0 +1,311 @@
+ * Copyright (c) 2019 Google, Inc.  All rights reserved.
+ * *******************************************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/*
+ * memquery_test.h - unit test for unix memquery logic, it is meant to be
+ * included into memquery.c after its includes.
+ *
+ * See also memquery_test_data.h
+ */
+
+#define MEMQUERY_ITERATOR_START memquery_iterator_start_test
+#define MEMQUERY_ITERATOR_NEXT memquery_iterator_next_test
+#define MEMQUERY_ITERATOR_STOP memquery_iterator_stop_test
+#define MODULE_IS_HEADER module_is_header_test
+#define MODULE_WALK_PROGRAM_HEADERS module_walk_program_headers_test
+
+typedef struct {
+    memquery_iter_t iter_result;
+    bool is_header;
+    app_pc mod_base, mod_end;
+} fake_memquery_result;
+
+typedef struct {
+    fake_memquery_result *iters;
+    size_t iters_count;
+
+    const char *in_name;
+    app_pc in_start;
+
+    int want_return;
+    app_pc want_start;
+    app_pc want_end;
+    const char *want_fulldir;
+    const char *want_filename;
+} memquery_library_bounds_test;
+
+#include "memquery_test_data.h"
+
+memquery_library_bounds_test cur_bounds;
+int next_iter_position = -1;
+bool enable_memquery_unit_test;
+
+/* CHECK_EQ verifies that the provded x and y are equal; if not, it prints a
+ * message to STDERR and returns false.
+ */
+#define CHECK_EQ(x, y, fmt)                                                  \
```

Comment:
Yeah I liked my version because it printed out more debug info for non-release builds.

I think you suggested that I move this to utils.h, and rename it to be EXPECT_PRINT? I've done that, let me know if I misinterpretted.

Reply:
Actually, the reason I decided to leave them in my function was because the existing EXPECT_* functions terminate the process on failure, whereas mine simply return false (so that the test can keep going, in my case). I figure having those two different functionalities named so similarly may be confusing; wdyt?

I don't /need/ to have them return false, I just happen to think it's convenient to not crash unit tests on the first error they find. So, instead of coming up with these special helpers I could just use the existing EXPECT function and make another one (which still exits the test early) for strings.

Reply:
The unit_tests test succeeds or fails based on exit code IIRC, not on golden output or anything, so in general these EXPECT checks have to kill the process in order the fail the test: some output is not going to do it.  To do otherwise would have to assume the calling code is going to do something like keep a count of failures and kill the process later which gets a little complex and fragile to assume unless we added some global counter and some shared exit check or sthg.  But  long-term for regression testing exiting on the first error is fine, right?  I assume it's only non-ideal when debugging the test itself?

Reply:
Correct, it is mostly about when we are debugging a failure; normal circumstances does not require anything more fancy than just exiting on the first error.

I did actually write support in the memquery unit test for this, but I'll just take it out to keep things simple.